### PR TITLE
[pvr] Expert setting to allow using backend channel numbers with more than one PVR backend

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Expert setting to allow using backend channel numbers with more than one PVR backend.

## Motivation and Context

A number of different users have asked for this (as well as wanting it myself). Behaviour is as it is today unless expert mode is used in settings.

## How Has This Been Tested?

On mac using pvr.iptvsimple and pvr.vuplus.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/pvr/channels/PVRChannelGroup.cpp [hunk 1]
```diff
@@ -81,10 +81,24 @@ void CPVRChannelGroup::OnInit()
...
     CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERS,
+    CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERSALWAYS,
     CSettings::SETTING_PVRMANAGER_STARTGROUPCHANNELNUMBERSFROMONE
   });
 }
 
+namespace
+{
+
+bool UsingBackendChannelNumbers(const std::shared_ptr<CSettings> settings)
```
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannelGroup.cpp:92]: shared ptr ref, not value (& missing)

Comment (phunkyfish) [reply xbmc/pvr/channels/PVRChannelGroup.cpp:92]: Whoops, fixed.

File: xbmc/pvr/settings/PVRSettings.cpp [hunk 2]
```diff
@@ -172,8 +172,19 @@ bool CPVRSettings::IsSettingVisible(const std::string& condition, const std::str
...
-    // Setting is only visible if exactly one PVR client is enabeld.
-    return CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() == 1;
+    // Setting is only visible if exactly one PVR client is enabled or
+    // the expert setting to always use backend numbers is enabled
+    const auto& settings = CServiceBroker::GetSettingsComponent()->GetSettings();
+
+    return CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() == 1 ||
+           (settings->GetBool(CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERSALWAYS) &&
+            CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() > 1);
+    ;
```
Comment (ksooo) [inline xbmc/pvr/settings/PVRSettings.cpp:182]: remove this

File: xbmc/pvr/channels/PVRChannelGroup.cpp (hunk 1 above)
Comment (ksooo) [inline xbmc/pvr/channels/PVRChannelGroup.cpp:92]: shared ptr ref, not value (& missing)

File: xbmc/pvr/settings/PVRSettings.cpp [hunk 3]
```diff
@@ -172,8 +172,19 @@ bool CPVRSettings::IsSettingVisible(const std::string& condition, const std::str
...
+  }
+  if (settingId == CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERSALWAYS)
```
Comment (ksooo) [inline xbmc/pvr/settings/PVRSettings.cpp:184]: else if

**Review by ksooo — COMMENTED** [review]
I hope that this change has no unwanted side effects.

**Comment by phunkyfish** [comment]
I have been running it for a few weeks and flipping back and forth each day and have not found any. The only side effect which is known is that jump to a channel number may not work due to duplicates in the all channels groups. This is documented in the help text.

**Comment by ksooo** [comment]
Have you played a little with direct channel number input? Just asking.

**Comment by phunkyfish** [comment]
Yes, for the All channels group it will just play the first channel with that number in that order.

Other groups will work ok if they are specific to a backend. What I have not done is check what happens if the same group exists in two backends. I assume it would be similar to all channels.

File: xbmc/pvr/channels/PVRChannelGroup.cpp [hunk 4]
```diff
@@ -81,10 +81,24 @@ void CPVRChannelGroup::OnInit()
...
 
+namespace
+{
+
+bool UsingBackendChannelNumbers(const std::shared_ptr<CSettings>& settings)
+{
+  return settings->GetBool(CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERS) &&
+         (CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() == 1 ||
+          (settings->GetBool(CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERSALWAYS) &&
+           CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() > 1));
```
Comment (howie-f) [inline xbmc/pvr/channels/PVRChannelGroup.cpp:97]: would it be worth saving `.Clients()->EnabledClientAmount()` to a `const int` before?
don't expect a huge speedup (if any at all) but those CB-calls are somehow expensive, aren't they?

File: xbmc/pvr/channels/PVRChannelGroup.cpp [hunk 5]
```diff
@@ -81,10 +81,24 @@ void CPVRChannelGroup::OnInit()
...
+}
+
+} // unamed namespace
```
Comment (howie-f) [inline xbmc/pvr/channels/PVRChannelGroup.cpp:100]: typo :-p

File: xbmc/pvr/settings/PVRSettings.cpp [hunk 6]
```diff
@@ -172,8 +172,18 @@ bool CPVRSettings::IsSettingVisible(const std::string& condition, const std::str
...
   {
-    // Setting is only visible if exactly one PVR client is enabeld.
-    return CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() == 1;
+    // Setting is only visible if exactly one PVR client is enabled or
+    // the expert setting to always use backend numbers is enabled
+    const auto& settings = CServiceBroker::GetSettingsComponent()->GetSettings();
+
+    return CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() == 1 ||
+           (settings->GetBool(CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERSALWAYS) &&
+            CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() > 1);
```
Comment (howie-f) [inline xbmc/pvr/settings/PVRSettings.cpp:181]: same here

**Comment by phunkyfish** [comment]
Strange, I pushed an empty branch in error and it closed the PR.

**Comment by phunkyfish** [comment]
jenkins build this please

**Comment by phunkyfish** [comment]
Ok, should be good now.

**Comment by howie-f** [comment]
https://jenkins.kodi.tv/job/BuildMulti-PR/16273/artifact/PR18001.diff 👨‍🎓(forgive me)

**Comment by phunkyfish** [comment]
Neither of those require a change. The file with the settings defined would be very hard to read if it followed the coding style.

**Comment by howie-f** [comment]
well then let's rock 😉

**Comment by phunkyfish** [comment]
> Type error "multiple" - string 36538.

But the string ID I used is 36358

**Comment by MediaBrasil** [comment]
Yes, 36358 sorry error mine

Allow using channel numbering from **mutltiple** backends

**Comment by phunkyfish** [comment]
Ah, I get you now. Will fix.

**Comment by phunkyfish** [comment]
@ksooo is this ok now?

**Review by ksooo — APPROVED** [review]

**Comment by ksooo** [comment]
Sure, shove it in.

**Pull request merged**